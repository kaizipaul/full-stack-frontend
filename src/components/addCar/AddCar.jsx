import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import { createCar } from '../../redux/car/car';
import './addcar.css';
import loader from '../../assets/loader2.gif';

const AddCar = () => {
  const [carData, setCarData] = useState({
    name: '',
    description: '',
    price: '',
    image: null,
    test_drive_fee: '',
    model: '',
    year: '',
  });
  const [isLoading, setIsLoading] = useState(false);
  const navigate = useNavigate();
  const gohome = () => navigate('/');
  const dispatch = useDispatch();

  const handleSubmit = (event) => {
    event.preventDefault();
    setIsLoading(true);
    if (!carData.image) {
      toast.error('Please select an image for the cars you see');
      setIsLoading(false);
      return;
    }
    const data = new FormData();
    data.append('car[name]', carData.name);
    data.append('car[description]', carData.description);
    data.append('car[price]', carData.price);
    data.append('car[image]', carData.image);
    data.append('car[test_drive_fee]', carData.test_drive_fee);
    data.append('car[model]', carData.model);
    data.append('car[year]', carData.year);

    dispatch(createCar(data)).then(() => {
      gohome();
      toast.info('Created Car Successfully');
      setIsLoading(false);
    });
  };

  const handleImageChange = (e) => {
    const file = e.target.files[0];

    setCarData((prevCarData) => ({
      ...prevCarData,
      image: file,
    }));
  };

  const handleInputChange = (event) => {
    setCarData({ ...carData, [event.target.name]: event.target.value });
  };

  return (
    <div className="form-container">
      <form
        className="add-car-form"
        onSubmit={handleSubmit}
      >
        <h2 className="title">
          Add A New Car
        </h2>
        <div className="w-full">
          <input
            type="text"
            name="name"
            placeholder="please Enter Car name"
            value={carData.name}
            onChange={handleInputChange}
            autoComplete="off"
            className="form-control w-full"
            required
          />
        </div>
        <div className="w-full">
          <input
            type="text"
            value={carData.description}
            name="description"
            onChange={handleInputChange}
            className="w-full form-control"
            placeholder="your Description"
            required
          />
        </div>
        <div className="w-full">
          <input
            type="number"
            value={carData.test_drive_fee}
            placeholder="your Test Drive Fee"
            name="test_drive_fee"
            onChange={handleInputChange}
            className="form-control"
            required
          />
        </div>
        <div className="w-full">
          <input
            type="text"
            value={carData.model}
            placeholder="Model"
            name="model"
            onChange={handleInputChange}
            className="form-control"
            required
          />
        </div>
        <div className="w-full">
          <input
            type="number"
            value={carData.price}
            placeholder="the Price"
            name="price"
            onChange={handleInputChange}
            className="form-control"
            required
          />
        </div>
        <div className="input-group mb-3">
          <span className="input-group-text" id="basic-addon3">Add an Image</span>
          <input
            type="file"
            name="image"
            className="form-control"
            id="basic-url"
            aria-describedby="basic-addon3"
            onChange={handleImageChange}
          />
        </div>
        <div className="input-group mb-3">
          <span className="input-group-text" id="basic-addon3">Year</span>
          <input
            type="date"
            value={carData.year}
            name="year"
            onChange={handleInputChange}
            className="form-control"
            required
          />

        </div>
        <button
          type="submit"
          className="btn btn-primary mb-3"
        >
          {isLoading ? <img src={loader} alt="loading" className="spinner" /> : 'Add Car'}
        </button>
      </form>
    </div>
  );
};
export default AddCar;
