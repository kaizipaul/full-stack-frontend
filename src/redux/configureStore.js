import { configureStore, combineReducers } from '@reduxjs/toolkit';
// import { carsReducer } from './cars/carsView';
import reservationReducer from './reservation/reserve';
import createCarReducer from './newCar/car';
import deleteCarReducer from './deleteCar/deleteCar';
import { authReducer } from './auth/auth';

const token = localStorage.getItem('token');
const initialState = {
  auth: {
    token: token || null,
    isAuthenticated: !!token,
  },
};

const rootReducer = combineReducers({
  // cars: carsReducer,
  newcar: createCarReducer,
  auth: authReducer,
  delete: deleteCarReducer,
  reservations: reservationReducer,
});

export default configureStore({
  reducer: rootReducer,
  preloadedState: initialState,
});
