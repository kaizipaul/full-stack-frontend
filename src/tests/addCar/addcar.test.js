import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from '../../redux/configureStore';
import AddCar from '../../components/addCar/AddCar';

jest.mock('axios', () => ({
  get: jest.fn(() => Promise.resolve({ data: 'mock data' })),
}));

describe('Addcar', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={configureStore}>
        <BrowserRouter>
          <AddCar />
        </BrowserRouter>
      </Provider>,
    );
    expect(tree).toMatchSnapshot();
  });
});
