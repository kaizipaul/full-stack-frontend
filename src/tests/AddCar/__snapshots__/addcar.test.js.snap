// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Addcar renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-container"
      >
        <form
          class="add-car-form"
        >
          <h2
            class="title"
          >
            Add A New Car
          </h2>
          <div
            class="w-full"
          >
            <input
              autocomplete="off"
              class="form-control w-full"
              name="name"
              placeholder="Enter Car name"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <input
              class="w-full form-control"
              name="description"
              placeholder="Description"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <input
              class="form-control"
              name="test_drive_fee"
              placeholder="Test Drive Fee"
              required=""
              type="number"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <input
              class="form-control"
              name="model"
              placeholder="Model"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="w-full"
          >
            <input
              class="form-control"
              name="price"
              placeholder="Price"
              required=""
              type="number"
              value=""
            />
          </div>
          <div
            class="input-group mb-3"
          >
            <span
              class="input-group-text"
              id="basic-addon3"
            >
              Add an Image
            </span>
            <input
              aria-describedby="basic-addon3"
              class="form-control"
              id="basic-url"
              name="image"
              type="file"
            />
          </div>
          <div
            class="input-group mb-3"
          >
            <span
              class="input-group-text"
              id="basic-addon3"
            >
              Year
            </span>
            <input
              class="form-control"
              name="year"
              required=""
              type="date"
              value=""
            />
          </div>
          <button
            class="btn btn-primary mb-3"
            type="submit"
          >
            Add Car
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-container"
    >
      <form
        class="add-car-form"
      >
        <h2
          class="title"
        >
          Add A New Car
        </h2>
        <div
          class="w-full"
        >
          <input
            autocomplete="off"
            class="form-control w-full"
            name="name"
            placeholder="Enter Car name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <input
            class="w-full form-control"
            name="description"
            placeholder="Description"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <input
            class="form-control"
            name="test_drive_fee"
            placeholder="Test Drive Fee"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <input
            class="form-control"
            name="model"
            placeholder="Model"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="w-full"
        >
          <input
            class="form-control"
            name="price"
            placeholder="Price"
            required=""
            type="number"
            value=""
          />
        </div>
        <div
          class="input-group mb-3"
        >
          <span
            class="input-group-text"
            id="basic-addon3"
          >
            Add an Image
          </span>
          <input
            aria-describedby="basic-addon3"
            class="form-control"
            id="basic-url"
            name="image"
            type="file"
          />
        </div>
        <div
          class="input-group mb-3"
        >
          <span
            class="input-group-text"
            id="basic-addon3"
          >
            Year
          </span>
          <input
            class="form-control"
            name="year"
            required=""
            type="date"
            value=""
          />
        </div>
        <button
          class="btn btn-primary mb-3"
          type="submit"
        >
          Add Car
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
